rhp/v4/rpc_test.go

```diff
@@ -765,6 +767,184 @@ func TestRPCRenew(t *testing.T) {
+		Contract: revision,
+		Target:   types.Siacoins(1),
+	}
+	var expectedCost types.Currency
+	for _, balance := range balances {
+		replenishParams.Accounts = append(replenishParams.Accounts, balance.Account)
+		if balance.Balance.Cmp(replenishParams.Target) < 0 {
+			expectedCost = expectedCost.Add(replenishParams.Target.Sub(balance.Balance))
+		}
+	}
+
+	// replenish the accounts
+	replenishResult, err := rhp4.RPCReplenishAccounts(context.Background(), transport, replenishParams, cs, fundAndSign)
+	if err != nil {
+		t.Fatal(err)
+	} else if !replenishResult.Usage.AccountFunding.Equals(expectedCost) {
+		t.Fatalf("expected %v, got %v", expectedCost, replenishResult.Usage.AccountFunding)
+	} else if len(replenishResult.Deposits) != len(replenishParams.Accounts) {
+		t.Fatalf("expected %v, got %v", len(replenishParams.Accounts), len(replenishResult.Deposits))
+	}
+	revisionTransfer := revision.Revision.RenterOutput.Value.Sub(replenishResult.Revision.RenterOutput.Value)
+	if !revisionTransfer.Equals(replenishResult.Usage.AccountFunding) {
+		t.Fatalf("expected %v, got %v", replenishResult.Usage.AccountFunding, revisionTransfer)
+	}
+	revision.Revision = replenishResult.Revision
+
+	balanceMap := make(map[proto4.Account]types.Currency)
+	for _, account := range balances {
+		balance, err := rhp4.RPCAccountBalance(context.Background(), transport, account.Account)
+		if err != nil {
+			t.Fatal(err)
+		}
+		balanceMap[account.Account] = balance
+		if account.Balance.Cmp(replenishParams.Target) < 0 {
+			if !balance.Equals(replenishParams.Target) {
+				t.Fatalf("expected %v, got %v", replenishParams.Target, balance)
+			}
+		} else if !balance.Equals(account.Balance) {
+			// balances that were already >= 1SC should not have changed
+			t.Fatalf("expected %v, got %v", account.Balance, balance)
+		}
+	}
+
+	// call replenish again with the same accounts, the balances should not change
+	replenishParams.Contract = revision
+	replenishResult, err = rhp4.RPCReplenishAccounts(context.Background(), transport, replenishParams, cs, fundAndSign)
+	if err != nil {
+		t.Fatal(err)
+	} else if !replenishResult.Usage.AccountFunding.Equals(types.ZeroCurrency) {
+		t.Fatalf("expected %v, got %v", expectedCost, replenishResult.Usage.AccountFunding)
+	} else if len(replenishResult.Deposits) != len(replenishParams.Accounts) {
+		t.Fatalf("expected %v, got %v", len(replenishParams.Accounts), len(replenishResult.Deposits))
+	}
+	for _, deposit := range replenishResult.Deposits {
+		if !deposit.Amount.IsZero() {
+			t.Fatalf("expected zero, got %v", deposit.Amount)
+		}
+	}
+
+	for _, account := range balances {
+		balance, err := rhp4.RPCAccountBalance(context.Background(), transport, account.Account)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		expected, ok := balanceMap[account.Account]
+		if !ok {
+			t.Fatalf("expected account not found")
+		} else if !balance.Equals(expected) {
+			t.Fatalf("expected %v, got %v", expected, balance)
+		}
+	}
+}
```

Comment:
Considering that we've had this issue in the past for I think renewals can we try and add that new revision to the transaction pool at the end here to make sure it's a valid revision for all parties?

Reply:
No, I still need to do it. Hold off on re-review please.

Reply:
I added a helper and the check to everywhere we return a revision. I could move it to a follow-up if you don't like the diff, but it did expose another place we weren't setting the host sig on the renter side.